### PR TITLE
refactor(platform): replace bls library

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented
bls-signatures built from c++ had a lot of compilation issues for Windows. In this PR we replace it with https://crates.io/crates/blsful/3.0.0-pre6

## What was done?
Of note: the serialization for Platform version does change and needs to be checked, it should work though

## How Has This Been Tested?
Tests still pass

## Breaking Changes
Should not require a platform version change.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

## Release Notes

- **New Features**
	- Enhanced handling of BLS signatures and public keys across various modules, improving type safety and error handling.
	- Introduced specific BLS signature implementation (`Bls12381G2Impl`) for public key and signature types, ensuring consistency in cryptographic operations.
	- Added new methods for managing and converting between different versions of quorum sets.
	- Added a new public export for the `dapi_grpc` module in the SDK.
	- Updated the `run_block_proposal_v0` function to improve error handling and control flow.

- **Bug Fixes**
	- Improved error handling in signature verification processes and public key conversions.
	- Corrected a typographical error in the error message for `InvalidInstantAssetLockProofSignatureError`.

- **Documentation**
	- Updated comments and documentation to reflect changes in method signatures and functionality.

- **Chores**
	- Incremented dependency versions for `dashcore-rpc` across multiple packages.
	- Updated dependency versions for `lru` in the `dash-sdk` and `simple-signer` packages.

### CODE REVIEW

**Comment by lklimek** [comment]
Local devnet on BLS branch fails with:
```
errors=[Abci(InvalidChainLock("received a chain lock for height 1 that we figured out was invalid based on platform state ChainLock
```

File: packages/rs-dpp/src/identity/identity_public_key/methods/hash/mod.rs [hunk 1]
```diff
@@ -14,7 +14,7 @@ impl IdentityPublicKeyHashMethodsV0 for IdentityPublicKey {
 
     fn validate_private_key_bytes(
         &self,
-        private_key_bytes: &[u8],
+        private_key_bytes: &[u8; 32],
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/identity/identity_public_key/methods/hash/mod.rs:17]: Done.

File: packages/rs-dpp/src/core_types/validator_set/mod.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-use crate::bls_signatures::PublicKey as BlsPublicKey;
+use crate::bls_signatures::{Bls12381G2Impl, PublicKey as BlsPublicKey};
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/core_types/validator_set/mod.rs:1]: No need.

File: packages/rs-dpp/src/identity/random.rs [hunk 3]
```diff
@@ -66,8 +66,8 @@ impl Identity {
     ) -> Result<(Self, I), ProtocolError>
     where
         I: Default
-            + IntoIterator<Item = (IdentityPublicKey, Vec<u8>)>
-            + Extend<(IdentityPublicKey, Vec<u8>)>,
+            + IntoIterator<Item = (IdentityPublicKey, [u8; 32])>
+            + Extend<(IdentityPublicKey, [u8; 32])>,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/identity/random.rs:70]: Must have been previously fixed.

File: packages/rs-dpp/src/bls/native_bls.rs [hunk 4]
```diff
@@ -18,31 +20,47 @@ impl BlsModule for NativeBlsModule {
...
+        let g2_element =
+            <Bls12381G2Impl as Pairing>::Signature::from_compressed(&signature_96_bytes)
+                .into_option()
+                .ok_or(anyhow!("signature derivation failed"))?;
+
+        let signature = Signature::Basic(g2_element);
+
+        match signature.verify(&public_key, data) {
+            Ok(_) => Ok(true),
+            Err(_) => Err(anyhow!("Verification failed").into()),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/bls/native_bls.rs:37]: fixed

File: packages/rs-dpp/src/bls/native_bls.rs [hunk 5]
```diff
@@ -18,31 +20,47 @@
...
             .try_into()
             .map_err(|_| anyhow!("the BLS private key must be 32 bytes long"))?;
-        let pk = PrivateKey::from_bytes(&fixed_len_key, false).map_err(anyhow::Error::msg)?;
-        let public_key = pk.g1_element().map_err(anyhow::Error::msg)?;
-        let public_key_bytes = public_key.to_bytes().to_vec();
+        let pk = SecretKey::<Bls12381G2Impl>::from_be_bytes(&fixed_len_key)
+            .into_option()
+            .ok_or(anyhow!("Incorrect Priv Key"))?;
+        let public_key = pk.public_key();
+        let public_key_bytes = public_key.0.to_compressed().to_vec();
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/bls/native_bls.rs:49]: This is not my library.

File: packages/rs-dpp/src/bls/native_bls.rs [hunk 6]
```diff
@@ -18,31 +20,47 @@
...
     fn private_key_to_public_key(&self, private_key: &[u8]) -> Result<Vec<u8>, ProtocolError> {
         let fixed_len_key: [u8; 32] = private_key
             .try_into()
             .map_err(|_| anyhow!("the BLS private key must be 32 bytes long"))?;
-        let pk = PrivateKey::from_bytes(&fixed_len_key, false).map_err(anyhow::Error::msg)?;
-        let public_key = pk.g1_element().map_err(anyhow::Error::msg)?;
-        let public_key_bytes = public_key.to_bytes().to_vec();
+        let pk = SecretKey::<Bls12381G2Impl>::from_be_bytes(&fixed_len_key)
+            .into_option()
+            .ok_or(anyhow!("Incorrect Priv Key"))?;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/bls/native_bls.rs:47]: Fixed

File: packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving_v1.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+use dpp::bls_signatures::Bls12381G2Impl;
+use dpp::identity::state_transition::asset_lock_proof::Encode;
+use dpp::platform_serialization::de::Decode;
+use dpp::platform_value::Bytes32;
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct SignatureVerificationQuorumSetForSavingV1 {
+    config: QuorumConfigForSavingV0,
+    current_quorums: Vec<QuorumForSavingV1>,
+    previous_quorums: Option<PreviousPastQuorumsForSavingV0>,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving_v1.rs:19]: intentional.

File: packages/rs-drive-abci/src/platform_types/platform_state/v0/old_structures/mod.rs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        Self {
+            quorum_hash,
+            quorum_index,
+            core_height,
+            members: members
+                .into_iter()
+                .map(|(pro_tx_hash, validator)| (pro_tx_hash, validator.into()))
+                .collect(),
+            threshold_public_key: PublicKey::try_from(threshold_public_key.to_bytes().as_slice())
+                .unwrap(),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/platform_state/v0/old_structures/mod.rs:54]: I used an expect instead.

File: packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/mod.rs [hunk 9]
```diff
@@ -4,6 +4,7 @@ use crate::config::QuorumLikeConfig;
 use crate::error::execution::ExecutionError;
 use crate::error::Error;
 use crate::platform_types::signature_verification_quorum_set::v0::for_saving::SignatureVerificationQuorumSetForSavingV0;
+use crate::platform_types::signature_verification_quorum_set::v0::for_saving_v1::SignatureVerificationQuorumSetForSavingV1;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/mod.rs:7]: this is okay.

File: packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/mod.rs [hunk 10]
```diff
@@ -118,13 +119,15 @@
 pub enum SignatureVerificationQuorumSetForSaving {
     /// Version 0 of the signature verification quorums
     V0(SignatureVerificationQuorumSetForSavingV0),
+    /// Version 0 of the signature verification quorums
+    V1(SignatureVerificationQuorumSetForSavingV1),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/mod.rs:123]: Fixed

File: packages/rs-dpp/src/signing.rs [hunk 11]
```diff
@@ -38,25 +40,34 @@
...
-                            SignatureError::BasicBLSError(BasicBLSError::new(e.to_string())).into(),
-                        );
+                            SignatureError::BasicBLSError(BasicBLSError::new(format!(
+                                "Signature was {} bytes, expected 96 bytes",
+                                signature.len()
+                            )))
+                            .into(),
+                        )
                     }
                 };
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/signing.rs:69]: this is okay.

File: packages/rs-drive-abci/src/platform_types/commit/v0/mod.rs [hunk 12]
```diff
@@ -120,11 +124,11 @@ impl CommitV0 {
...
 
-        match public_key.verify(&signature, &hash) {
-            true => ValidationResult::default(),
-            false => ValidationResult::new_with_error(AbciError::BadCommitSignature(format!(
+        match signature.verify(&public_key, &hash) {
+            Ok(_) => ValidationResult::default(),
+            Err(_) => ValidationResult::new_with_error(AbciError::BadCommitSignature(format!(
                 "commit signature {} is wrong",
-                hex::encode(signature.to_bytes().as_slice())
+                signature
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/commit/v0/mod.rs]: already had been fixed

File: packages/rs-dpp/src/signing.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/signing.rs:69]: nevermind, i took this suggestion.

File: packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving.rs [hunk 13]
```diff
@@ -159,7 +162,10 @@ impl Into<Vec<QuorumForSavingV0>> for Quorums<VerificationQuorum> {
         self.into_iter()
             .map(|(hash, quorum)| QuorumForSavingV0 {
                 hash: Bytes32::from(hash.as_byte_array()),
-                public_key: quorum.public_key,
+                public_key: bls_signatures::PublicKey::from_bytes(
+                    &quorum.public_key.0.to_compressed(),
+                )
+                .expect("expected to convert between BLS key libraries (to chia)"),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving.rs:168]: This is something that happens once and can not really fail, expect is fine.

File: packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving.rs [hunk 14]
```diff
@@ -145,7 +145,10 @@
             (
                 QuorumHash::from_byte_array(quorum.hash.to_buffer()),
                 VerificationQuorum {
-                    public_key: quorum.public_key,
+                    public_key: dpp::bls_signatures::PublicKey::try_from(
+                        quorum.public_key.to_bytes().as_slice(),
+                    )
+                    .expect("expected to convert between BLS key libraries (from chia)"),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/signature_verification_quorum_set/v0/for_saving.rs:151]: This is fine, as we already had verified it was correct.

File: packages/rs-drive-abci/tests/strategy_tests/masternode_list_item_helpers.rs [hunk 15]
```diff
@@ -32,13 +33,16 @@ impl UpdateMasternodeListItem for MasternodeListItem {
...
+                    let private_key_operator_bytes = bls_signatures::PrivateKey::generate_dash(rng)
+                        .expect("expected to generate a private key")
                         .to_bytes()
                         .to_vec();
+                    let private_key_operator = BlsPrivateKey::<Bls12381G2Impl>::from_be_bytes(
+                        &private_key_operator_bytes.try_into().unwrap(),
+                    )
+                    .unwrap();
+                    let pub_key_operator =
+                        private_key_operator.public_key().0.to_compressed().to_vec();
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/tests/strategy_tests/masternode_list_item_helpers.rs:45]: fixed

File: packages/rs-drive-abci/src/mimic/test_quorum.rs [hunk 16]
```diff
@@ -3,8 +3,9 @@ use crate::platform_types::validator_set::v0::ValidatorSetV0;
 use dashcore_rpc::dashcore::hashes::Hash;
 use dashcore_rpc::dashcore::{ProTxHash, PubkeyHash, QuorumHash};
 use dashcore_rpc::dashcore_rpc_json::{QuorumInfoResult, QuorumMember, QuorumType};
-use dpp::bls_signatures;
-use dpp::bls_signatures::{PrivateKey as BlsPrivateKey, PublicKey as BlsPublicKey};
+use dpp::bls_signatures::{
+    Bls12381G2Impl, PublicKey as BlsPublicKey, PublicKey, SecretKey as BlsPrivateKey, SecretKey,
+};
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/mimic/test_quorum.rs:8]: it's fine

File: packages/rs-drive-abci/src/mimic/test_quorum.rs [hunk 17]
```diff
@@ -154,8 +156,12 @@
...
                     pro_tx_hash: ProTxHash::from_slice(pro_tx_hash.as_slice())
                         .expect("expected 32 bytes for pro_tx_hash"),
-                    private_key: key,
-                    public_key,
+                    private_key: SecretKey::from_be_bytes(
+                        &key.to_bytes().to_vec().try_into().expect("32 bytes"),
+                    )
+                    .expect("expected conversion to work"),
+                    public_key: PublicKey::try_from(public_key.to_bytes().as_slice())
+                        .expect("expected conversion to work"),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/mimic/test_quorum.rs:164]: fixed

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/mimic/test_quorum.rs:164]: no, this is fine

File: packages/rs-drive-abci/src/mimic/test_quorum.rs [hunk 18]
```diff
@@ -255,7 +269,7 @@
                     pro_tx_hash: *pro_tx_hash,
                     pub_key_operator: vec![], //doesn't matter
                     valid: true,
-                    pub_key_share: Some(public_key.to_bytes().to_vec()),
+                    pub_key_share: Some(public_key.0.to_compressed().to_vec()),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/mimic/test_quorum.rs:272]: Already done.

File: packages/rs-drive-abci/src/mimic/test_quorum.rs [hunk 19]
```diff
@@ -178,8 +184,16 @@
...
+            private_key: SecretKey::from_be_bytes(
+                &recovered_private_key
+                    .to_bytes()
+                    .to_vec()
+                    .try_into()
+                    .expect("32 bytes"),
+            )
+            .expect("expected conversion to work"),
+            public_key: PublicKey::try_from(public_key.to_bytes().as_slice())
+                .expect("expected conversion to work"),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/mimic/test_quorum.rs:196]: this is fine.

File: packages/rs-dpp/src/identity/identity_public_key/v0/methods/mod.rs [hunk 20]
```diff
@@ -166,7 +205,7 @@ mod tests {
         );
 
         // Test with an invalid private key
-        let invalid_private_key_bytes = vec![0u8; private_key_data.len()];
+        let invalid_private_key_bytes = [0u8; 32];
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/identity/identity_public_key/v0/methods/mod.rs:203]: already done.

File: packages/rs-drive-abci/src/platform_types/validator_set/v0/mod.rs [hunk 21]
```diff
@@ -99,7 +99,9 @@ impl ValidatorSetMethodsV0 for ValidatorSetV0 {
                                 Some(Ok(abci::ValidatorUpdate {
                                     pub_key: public_key.clone().map(|public_key| {
                                         crypto::PublicKey {
-                                            sum: Some(Bls12381(public_key.to_bytes().to_vec())),
+                                            sum: Some(Bls12381(
+                                                public_key.0.to_compressed().to_vec(),
+                                            )),
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/platform_types/validator_set/v0/mod.rs:104]: This is fine.

File: packages/rs-dpp/src/core_types/validator_set/v0/mod.rs [hunk 22]
```diff
@@ -85,7 +86,7 @@ impl Encode for ValidatorSetV0 {
 
         // Custom encoding for BlsPublicKey if needed
         // Assuming BlsPublicKey can be serialized to a byte slice
-        let public_key_bytes = *self.threshold_public_key.to_bytes();
+        let public_key_bytes = self.threshold_public_key.0.to_compressed();
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/core_types/validator_set/v0/mod.rs:89]: already done.

File: packages/rs-dpp/src/core_types/validator_set/v0/mod.rs [hunk 23]
```diff
@@ -61,7 +62,7 @@
                     pro_tx_hash, validator.node_ip
                 ))
                 .join(", "),
-            hex::encode(self.threshold_public_key.to_bytes().as_slice()) // Assuming BlsPublicKey is a byte array
+            hex::encode(self.threshold_public_key.0.to_compressed()) // Assuming BlsPublicKey is a byte array
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/core_types/validator_set/v0/mod.rs:65]: already done.

File: packages/rs-dpp/src/core_types/validator_set/v0/mod.rs [hunk 24]
```diff
@@ -162,11 +166,12 @@
...
-            bincode::error::DecodeError::OtherString(
-                "Failed to decode BlsPublicKey in borrow decode".to_string(),
-            )
-        })?;
+        let threshold_public_key =
+            BlsPublicKey::try_from(public_key_bytes.as_slice()).map_err(|_| {
+                bincode::error::DecodeError::OtherString(
+                    "Failed to decode BlsPublicKey in borrow decode".to_string(),
+                )
+            })?;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/core_types/validator_set/v0/mod.rs:174]: Done.

File: packages/rs-dpp/src/core_types/validator_set/v0/mod.rs [hunk 25]
```diff
@@ -118,9 +119,12 @@
...
         public_key_bytes.copy_from_slice(&bytes);
-        let threshold_public_key = BlsPublicKey::from_bytes(&public_key_bytes).map_err(|_| {
-            bincode::error::DecodeError::OtherString("Failed to decode BlsPublicKey".to_string())
-        })?;
+        let threshold_public_key =
+            BlsPublicKey::try_from(public_key_bytes.as_slice()).map_err(|_| {
+                bincode::error::DecodeError::OtherString(
+                    "Failed to decode BlsPublicKey".to_string(),
+                )
+            })?;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/core_types/validator_set/v0/mod.rs:127]: Done.

File: packages/rs-dpp/src/identity/identity_public_key/key_type.rs [hunk 26]
```diff
@@ -236,7 +236,7 @@
             }
             KeyType::ECDSA_HASH160 => {
                 let secp = Secp256k1::new();
-                let secret_key = dashcore::secp256k1::SecretKey::from_slice(private_key_bytes)
+                let secret_key = dashcore::secp256k1::SecretKey::from_byte_array(private_key_bytes)
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/identity/identity_public_key/key_type.rs:240]: already done.

File: packages/rs-drive-abci/tests/strategy_tests/query.rs [hunk 27]
```diff
@@ -68,7 +69,7 @@ pub struct ProofVerification<'a> {
     pub signature: &'a [u8; 96],
 
     /// Threshold key used to verify the signature
-    pub public_key: &'a dpp::bls_signatures::PublicKey,
+    pub public_key: &'a dpp::bls_signatures::PublicKey<Bls12381G2Impl>,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/tests/strategy_tests/query.rs:72]: already done.

File: packages/rs-dpp/src/identity/identity_public_key/key_type.rs [hunk 28]
```diff
@@ -204,13 +200,13 @@
...
         &self,
-        private_key_bytes: &[u8],
+        private_key_bytes: &[u8; 32],
         network: Network,
     ) -> Result<Vec<u8>, ProtocolError> {
         match self {
             KeyType::ECDSA_SECP256K1 => {
                 let secp = Secp256k1::new();
-                let secret_key = dashcore::secp256k1::SecretKey::from_slice(private_key_bytes)
+                let secret_key = dashcore::secp256k1::SecretKey::from_byte_array(private_key_bytes)
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/identity/identity_public_key/key_type.rs:210]: already done.

File: packages/rs-drive-abci/tests/strategy_tests/masternodes.rs [hunk 29]
```diff
@@ -209,13 +209,15 @@ pub fn generate_test_masternodes(
...
-            .expect("expected to get public key")
+        let private_key_operator_bytes = bls_signatures::PrivateKey::generate_dash(rng)
+            .expect("expected to generate a private key")
             .to_bytes()
             .to_vec();
+        let private_key_operator = BlsPrivateKey::<Bls12381G2Impl>::from_be_bytes(
+            &private_key_operator_bytes.try_into().unwrap(),
+        )
+        .unwrap();
+        let pub_key_operator = private_key_operator.public_key().0.to_compressed().to_vec();
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/tests/strategy_tests/masternodes.rs:220]: this is done on purpose

**Review by shumkov — CHANGES_REQUESTED** [review]

**Review by shumkov — CHANGES_REQUESTED** [review]
Change base to v1.8-dev

File: packages/rs-dpp/src/bls/native_bls.rs [hunk 30]
```diff
@@ -1,12 +1,14 @@
+use crate::bls_signatures::{
+    Bls12381G2Impl, Pairing, PublicKey, SecretKey, Signature, SignatureSchemes,
+};
 use crate::{BlsModule, ProtocolError, PublicKeyValidationError};
-use anyhow::anyhow;
-use dashcore::bls_signatures::{self, PrivateKey, PublicKey};
+use std::array::TryFromSliceError;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/bls/native_bls.rs:5]: Fixed.

File: packages/rs-dpp/src/errors/protocol_error.rs [hunk 31]
```diff
@@ -1,3 +1,4 @@
+use std::array::TryFromSliceError;
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-dpp/src/errors/protocol_error.rs:1]: Fixed

**Pull request merged**